apps/events/models.py

```diff
@@ -0,0 +1,171 @@
+    NEW_BOT_MESSAGE = ("new_bot_message", "A new bot message is received")
+
+
+class StaticTrigger(BaseModel):
+    action = models.OneToOneField(EventAction, on_delete=models.CASCADE, related_name="static_trigger")
+    experiment = models.ForeignKey(Experiment, on_delete=models.CASCADE, related_name="static_triggers")
+    type = models.CharField(choices=StaticTriggerType.choices, db_index=True)
+    event_logs = GenericRelation(EventLog)
+
+    def fire(self, session):
+        try:
+            result = ACTION_FUNCTIONS[self.action.action_type](session, self.action.params)
+            # TODO: handle message here.
+            self.event_logs.create(session=session, status=EventLogStatusChoices.SUCCESS)
+            return result
+        except Exception:
+            # TODO: log exception reason in the event_log
+            self.event_logs.create(session=session, status=EventLogStatusChoices.FAILURE)
+
+
+class TimeoutTrigger(BaseModel):
+    action = models.OneToOneField(EventAction, on_delete=models.CASCADE, related_name="timeout_trigger")
+    experiment = models.ForeignKey(Experiment, on_delete=models.CASCADE, related_name="timeout_triggers")
+    delay = models.PositiveIntegerField(
+        help_text="Seconds to wait after last response before triggering action",
+    )
+    total_num_triggers = models.IntegerField(
+        default=1,
+        help_text="The number of times to trigger the action",
+    )
+    event_logs = GenericRelation(EventLog)
+
+    def timed_out_sessions(self):
+        """Finds all the timed out sessions where:
+        - The last human message was sent at a time earlier than the trigger time
+        - There have been fewer trigger attempts than the total number defined by the trigger
+        """
+
+        trigger_time = timezone.now() - timedelta(seconds=self.delay)
+
+        last_human_message_created_at = (
+            ChatMessage.objects.filter(
+                chat__experiment_session=OuterRef("pk"),
+                message_type=ChatMessageType.HUMAN,
+            )
+            .order_by("-created_at")
+            .values("created_at")[:1]
+        )
+        last_human_message_id = (
+            ChatMessage.objects.filter(
+                chat__experiment_session=OuterRef("session_id"),
+                message_type=ChatMessageType.HUMAN,
+            )
+            .order_by("-created_at")
+            .values("id")[:1]
+        )
+        log_count_for_last_message = (
+            EventLog.objects.filter(
+                session=OuterRef("pk"),
+                chat_message_id=Subquery(last_human_message_id),
+                status=EventLogStatusChoices.SUCCESS,
+            )
+            .annotate(
+                count=Func(F("chat_message_id"), function="Count")
+            )  # We don't use Count here because otherwise Django wants to do a group_by, which messes up the subquery: https://stackoverflow.com/a/69031027
+            .values("count")
+        )
+
+        sessions = (
```

Comment:
Perhaps for a future PR but I think if we want to replace the 'no activity pings' with this we'll need to be able to fire only for session where the last message was from a bot i.e. waiting for human response.

Reply:
Sounds good. Out of curiosity, would there ever be a time the last message _wasn't_ from a bot? Doesn't the bot always have the last word?

Reply:
That probably is true